### PR TITLE
Resolve index key length issues with index drops

### LINKED ISSUE #462: Schema: Define max index length for `user_id_action`
See https://wordpress.slack.com/archives/glotpress/p1465028403001050.

`KEY user_id_action (user_id,action)` 
should be
`KEY user_id_action (user_id,action($max_index_length))`

### LINKED ISSUE #464: Schema: Change field type for user IDs to `bigint(20)`
GlotPress stores IDs as `int(10)` but WordPress uses `bigint(20)`.

### PR DESCRIPTION
**Pull request by toolstack**
Alternate to #466 which drops the indexes before running `dbDelta()`;

This patch does several things:
- Converts all user_id's to bigint(20) to match WordPress
- Enforces a maximum size of components for compound indexes which would otherwise be too long
- Standardizes the 'status' field to be a varchar(20) in all tables
- Sets the size of the meta_key column in the meta table to be 255 to match WordPress
- Limits the size of the 'action' column in the permissions table to be 20 characters long
- Drops the changed indexes before running `dbDelta()`

Resolves #462.
Resolves #464.

### CODE REVIEW

File: gp-includes/install-upgrade.php [hunk 1]
```diff
@@ -12,9 +12,18 @@
...
-	dbDelta( implode( "\n", gp_schema_get() ) );
+	global $wpdb;
 
 	$gp_db_version = get_option( 'gp_db_version' );
+
+	if ( 970 < $gp_db_version ) {
+		$wpdb->query( "ALTER TABLE {$wpdb->translation_sets} DROP INDEX project_id_slug_locale, DROP INDEX locale_slug;" );
+		$wpdb->query( "ALTER TABLE {$wpdb->originals} DROP INDEX singular_plural_context;" );
+		$wpdb->query( "ALTER TABLE {$wpdb->meta} DROP INDEX object_type__meta_key, DROP INDEX object_type__object_id__meta_key;" );
+	}
```
Comment (ocean90) [inline gp-includes/install-upgrade.php:23]: The table names are missing the `gp_` prefix.

File: gp-includes/schema.php [hunk 2]
```diff
@@ -15,18 +15,25 @@ function gp_schema_get() {
...
+	 * concerned about that.
+	 *
+	 * As of WordPress 4.2, however, utf8mb4 is now the default collation which uses 4 bytes per character. This means that
+	 * an index which used to have room for floor(767/3) = 255 characters, now only has room for floor(767/4) = 191 characters.
+	 *
+	 * See https://dev.mysql.com/doc/refman/5.7/en/storage-requirements.html for details on the storage requirements
+	 * for each data type.
 	 */
-	$max_index_length = 191;
+	$max_index_charcters = 767;
```
Comment (ocean90) [inline gp-includes/schema.php]: Typo: `charcter(s)` => `character(s)`

File: gp-includes/schema.php [hunk 3]
```diff
@@ -145,24 +178,41 @@ function gp_schema_get() {
...
 
 	/*
 	 * Permissions
 	 */
 	$gp_schema['permissions'] = "CREATE TABLE $wpdb->gp_permissions (
 		id int(10) NOT NULL AUTO_INCREMENT,
-		user_id int(10) DEFAULT NULL,
-		action varchar(255) DEFAULT NULL,
+		user_id bigint(20) DEFAULT NULL,
+		action varchar(20) DEFAULT NULL,
```
Comment (ocean90) [inline gp-includes/schema.php]: I don't think we should change that, 20 seems too low for custom permission actions.

File: gp-includes/install-upgrade.php (hunk 1 above)
Comment (toolstack) [inline gp-includes/install-upgrade.php:23]: Fixed.

File: gp-includes/schema.php (hunk 2 above)
Comment (toolstack) [inline gp-includes/schema.php]: At least I was consistent in the typo ;)

Fixed.

File: gp-includes/schema.php (hunk 3 above)
Comment (toolstack) [inline gp-includes/schema.php]: I chose 20 as we use that length for the status field, but I'm not tied to it by any means.

I looked through our current actions at I believe the longest one was 9.  255 seemed excessive.  We could go somewhere in between 64 or 128 would both be reasonable sizes and not cause an index issue.

File: gp-includes/schema.php [hunk 4]
```diff
@@ -58,15 +65,32 @@ function gp_schema_get() {
...
 	 * Most projects will have only one translation set per locale.
 	 */
+
+	/*
+ 	 * The maximum length for the slug component of the project_id_slug_locale key is limited by the index size limit
+	 * minus the size of the project_id (4 bytes = 1 character) and locale (10 characters).
+	 *
+	 * Also make sure to never go over the length of the column.
+	 */
+	$max_pid_slug_locale_key_length = min( $max_index_length - 1 - 10, 255 );
```
Comment (ocean90) [inline gp-includes/schema.php:76]: > Also make sure to never go over the length of the column.

How could this happen? `$max_index_length` is 191 and always below 255.

Comment (toolstack) [inline gp-includes/schema.php:76]: This is a little bit of future proofing, I played with a version of the PR that detected which database type was being used (MyISAM, InnoDB, etc.) and allowed for longer key lengths (InnoDB can support up to 3000+ as a key size) which could cause the length of a component of a key to be longer than the column itself.

File: gp-includes/install-upgrade.php [hunk 5]
```diff
@@ -12,9 +12,18 @@
  * @since 1.0.0
  */
 function gp_upgrade_db() {
-	dbDelta( implode( "\n", gp_schema_get() ) );
+	global $wpdb;
 
 	$gp_db_version = get_option( 'gp_db_version' );
+
+	if ( 970 < $gp_db_version ) {
```
Comment (ocean90) [inline gp-includes/install-upgrade.php]: This should be `$gp_db_version && $gp_db_version < 970`
- It shouldn't run on installation.
- It should run if the db version is **below** 970.

File: gp-includes/schema.php (hunk 3 above)
Comment (ocean90) [inline gp-includes/schema.php]: Let's use 60 then.

**Pull request merged**